File: Kattis/unionfind.cpp
#include <bits/stdc++.h>
using namespace std;

class UnionFind
{
    struct UF { int p; int rank; };

    public:
        UnionFind(int n) {          // constructor
            howMany = n;
            uf = new UF[howMany];
            for (int i = 0; i < howMany; i++) {
                uf[i].p = i;
                uf[i].rank = 0;
            }
        }

        ~UnionFind() {
            delete[] uf;
        }

        int find(int x) { return find(uf,x); }        // for client use
      
        bool merge(int x, int y) {
            int res1, res2;
            res1 = find(uf, x);
            res2 = find(uf, y);
            if (res1 != res2) {
                if (uf[res1].rank > uf[res2].rank) {
                    uf[res2].p = res1;
                }
                else {
                    uf[res1].p = res2;
                    if (uf[res1].rank == uf[res2].rank) {
                        uf[res2].rank++;
                    }
                }
                return true;
            }
            return false;
        }
      
    private:
        int howMany;
        UF* uf;

        int find(UF uf[], int x) {     // recursive funcion for internal use
            if (uf[x].p != x) {
                uf[x].p = uf[uf[x].p].p;
                uf[x].p = find(uf, uf[x].p);
            }
            return uf[x].p;
        }
};

int main() {
    int n,q;
    cin>>n>>q;
    UnionFind uf = UnionFind(n);
    string ans = "";
    for(int i=0; i<q; i++) {
        char c;
        int a,b;
        cin>>c>>a>>b;
        
        if(c == '=') {
            uf.merge(a,b);
        } else {
            ans += ((uf.find(a) == uf.find(b)) ? "yes\n" : "no\n");
        }
    }
    cout<<ans<<endl;
}

File: Kattis/unionfind.py
class UnionFind:
    class UF:
        def __init__(self, p, rank):
            self.p = p
            self.rank = rank

    def __init__(self, n):
        self.size = n
        self.uf = []
        for i in range(self.size):
            self.uf.append(self.UF(i,0))


    def find(self, x):
        if self.uf[x].p != x:
            self.uf[x].p = self.find(self.uf[x].p)
        return self.uf[x].p

    def merge(self, u, v):
        a = self.find(u)
        b = self.find(v)
        if a != b:
            if self.uf[a].rank > self.uf[b].rank:
                self.uf[b].p = a
            else:
                self.uf[a].p = b
                if self.uf[a].rank == self.uf[b].rank:
                    self.uf[b].rank += 1
            return True
        return False

N, Q = [int(x) for x in input().split()]
uf = UnionFind(N)

ans = ""
for _ in range(Q):
    q,a,b = input().split()
    if q == "?":
        ans += "yes\n" if uf.find(int(a)) == uf.find(int(b)) else "no\n"
    else:
        uf.merge(int(a), int(b))

print(ans)
